proposals/patterns3.md

```diff
@@ -0,0 +1,273 @@
+1. If `P` is a type pattern, the *narrowed type* is the type of the type pattern's type.
+2. If `P` is a declaration pattern, the *narrowed type* is the type of the declaration pattern's type.
+3. If `P` is a recursive pattern that gives an explicit type, the *narrowed type* is that type.
+4. If `P` is a constant pattern where the constant is not the null constant and where the expression has no *constant expression conversion* to the *input type*, the *narrowed type* is the type of the constant.
+5. If `P` is a relational pattern where the constant expression has no *constant expression conversion* to the *input type*, the *narrowed type* is the type of the constant.
+6. If `P` is an `or` pattern, the *narrowed type* is the common type of the *narrowed type* of the subpatterns if such a common type exists. For this purpose, the common type algorithm considers only identity, boxing, and implicit reference conversions, and it considers all subpatterns of a sequence of `or` patterns (ignoring parenthesized patterns).
+7. If `P` is an `and` pattern, the *narrowed type* is the *narrowed type* of the right pattern. Moreover, the *narrowed type* of the left pattern is the *input type* of the right pattern.
+8. Otherwise the *narrowed type* of `P` is `P`'s input type.
+
+### Variable definitions and definite assignment
+
+The addition of `or` and `not` patterns creates some interesting new problems around pattern variables and definite assignment.  Since variables can normally be declared at most once, it would seem any pattern variable declared on one side of an `or` pattern would not be definitely assigned when the pattern matches.  Similarly, a variable declared inside a `not` pattern would not be expected to be definitely assigned when the pattern matches.  The simplest way to address this is to forbid declaring pattern variables in these contexts.  However, this may be too restrictive.  There are other approaches to consider.
+
+One scenario that is worth considering is this
+
+``` csharp
+if (e is not int i) return;
+M(i); // is i definitely assigned here?
+```
+
+This does not work today because, for an *is-pattern-expression*, the pattern variables are considered *definitely assigned* only where the *is-pattern-expression* is true ("definitely assigned when true").
+
+Supporting this would be simpler (from the programmer's perspective) than also adding support for a negated-condition `if` statement.  Even if we add such support, programmers would wonder why the above snippet does not work.  On the other hand, the same scenario in a `switch` makes less sense, as there is no corresponding point in the program where *definitely assigned when false* would be meaningful.  Would we permit this in an *is-pattern-expression* but not in other contexts where patterns are permitted?  That seems irregular.
+
```

Comment:
note (from our talk today): i think we should support this.  users will be able to say any of:

```
if (e is int)
if (e is not int)
if (e is int i)
// but not
if (e is not int i)
```

Despite that they *can* write:

```
if (!(e is int)) -> if (e is not int)
// but not
if (!(e is int i)) -> if (e is not int i)
```

It's a very strange inconsistency.

Reply:
We could do that but ONLY (1) for the is-pattern expression (no other places patterns could be used), and only (2) if the `not` is at the top level only. For other cases, there isn't any clear place that the pattern variables would be definitely assigned. If we did this, people would find these restrictions to be strange. I'm OK considering this relaxation as a separate proposal to be considered later.

Reply:
Agreed with Cyrus. It may make the spec convoluted, but I think the regularity would be worth it. We should definitely put this in the open issues tracker to make sure we don't forget about it.

Reply:
It is on the LDM agenda.